profile_test.go

```diff
@@ -0,0 +1,313 @@
+	//          687474703a2f2f6578616d706c652e636f6d # "http://example.com"
+	data := []byte{
+		0x72, 0x68, 0x74, 0x74, 0x70, 0x3a, 0x2f, 0x2f, 0x65,
+		0x78, 0x61, 0x6d, 0x70, 0x6c, 0x65, 0x2e, 0x63, 0x6f, 0x6d,
+	}
+
+	err = profile.UnmarshalCBOR(data)
+	assert.Nil(t, err)
+	recvURL, err := profile.Get()
+	assert.Nil(t, err)
+	assert.Equal(t, inputURL, recvURL)
+}
+
+// TestProfile_MarshalCBOR_OID tests the CBOR marshaling of profile set as OID
+func TestProfile_MarshalCBOR_OID(t *testing.T) {
+	inputOID := "2.5.2.8192"
+
+	profile, err := NewProfile(inputOID)
+	assert.Nil(t, err)
+	expected := []byte{
+		0x44, 0x55, 0x02, 0xC0, 0x00,
+	}
+	actual, err := profile.MarshalCBOR()
+	assert.Nil(t, err)
+	assert.Equal(t, expected, actual)
+
+	// Sample test from section 3.1: https://tools.ietf.org/html/draft-ietf-cbor-tags-oid-05#section-3
+	inputOID = "2.16.840.1.101.3.4.2.1"
+	expected = []byte{
+		0x49, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
+	}
+	err = profile.Set(inputOID)
+	assert.Nil(t, err)
+	actual, err = profile.MarshalCBOR()
+	assert.Nil(t, err)
+	assert.Equal(t, expected, actual)
+}
+
+//  TestProfile_UnMarshalCBOR_OID tests the CBOR unmarshaling of profile as decoded as OID
+func TestProfile_UnMarshalCBOR_OID(t *testing.T) {
+	inputOID := "2.5.2.8192"
+	profile, err := NewProfile(inputOID)
+	assert.Nil(t, err)
+
+	input := []byte{
+		0x44, 0x55, 0x02, 0xC0, 0x00,
+	}
+	err = profile.UnmarshalCBOR(input)
+	assert.Nil(t, err)
+	expectedOID, err := profile.Get()
+	assert.Nil(t, err)
+	assert.Equal(t, expectedOID, inputOID)
+
+	// Section 3.1 of https://tools.ietf.org/html/draft-ietf-cbor-tags-oid-05#section-3
+	inputOID = "2.16.840.1.101.3.4.2.1"
+
+	// CBOR Input
+	input = []byte{
+		0x49, 0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
+	}
+	err = profile.UnmarshalCBOR(input)
+	assert.Nil(t, err)
+	expectedOID, err = profile.Get()
+	assert.Nil(t, err)
+	assert.Equal(t, expectedOID, inputOID)
+}
+
+// TestProfile_MarshalJSON_URL tests the JSON Marshaling for a known URL
+func TestProfile_MarshalJSON_URL(t *testing.T) {
+	inputURL := "http://example.com"
+	profile, err := NewProfile(inputURL)
+	assert.Nil(t, err)
+
+	expected := []byte{
+		0x22, 0x68, 0x74, 0x74, 0x70, 0x3a, 0x2f, 0x2f, 0x65, 0x78,
+		0x61, 0x6d, 0x70, 0x6c, 0x65, 0x2e, 0x63, 0x6f, 0x6d, 0x22,
+	}
+
+	actual, err := profile.MarshalJSON()
+	assert.Nil(t, err)
+	assert.Equal(t, expected, actual)
+}
+
+// TestProfile_UnMarshalJSON_URL tests the UnMarshaling of a JSON value as URL string
+func TestProfile_UnMarshalJSON_URL(t *testing.T) {
+	inputURL := "http://example.com"
+	profile, err := NewProfile(inputURL)
+	assert.Nil(t, err)
+
+	data := []byte{
+		0x22, 0x68, 0x74, 0x74, 0x70, 0x3a, 0x2f, 0x2f, 0x65, 0x78,
+		0x61, 0x6d, 0x70, 0x6c, 0x65, 0x2e, 0x63, 0x6f, 0x6d, 0x22,
+	}
+
+	err = profile.UnmarshalJSON(data)
+	assert.Nil(t, err)
+	recvURL, err := profile.Get()
+	assert.Nil(t, err)
+	assert.Equal(t, inputURL, recvURL)
+
+	// Negative Test Cases
```

Comment:
this is a general comment, which I happen to stick here just because this bit is a good example of the point I'd like to make.  Units should be as granular as possible.   Ideally, we should test one interface at a time on a meaningful representation of possible inputs.  Besides, positive and negative conditions should be tested separately.

Reply:
Positive and Negative condition seperation done to completion!

Reply:
taken care of in #24